api4/websocket.go

```diff
@@ -19,30 +24,68 @@ func (api *API) InitWebSocket() {
 }
 
 func connectWebSocket(c *Context, w http.ResponseWriter, r *http.Request) {
-	fn := c.App.OriginChecker()
-	if fn != nil && !fn(r) {
-		c.Err = model.NewAppError("origin_check", "api.web_socket.connect.check_origin.app_error", nil, "", http.StatusBadRequest)
-		return
-	}
-
-	upgrader := ws.HTTPUpgrader{
-		Timeout: 5 * time.Second,
+	upgrader := websocket.Upgrader{
+		ReadBufferSize:  model.SOCKET_MAX_MESSAGE_SIZE_KB,
+		WriteBufferSize: model.SOCKET_MAX_MESSAGE_SIZE_KB,
+		CheckOrigin:     c.App.OriginChecker(),
 	}
 
-	conn, _, _, err := upgrader.Upgrade(r, w)
+	ws, err := upgrader.Upgrade(w, r, nil)
 	if err != nil {
 		c.Err = model.NewAppError("connect", "api.web_socket.connect.upgrade.app_error", nil, err.Error(), http.StatusInternalServerError)
 		return
 	}
 
-	wc := c.App.NewWebConn(conn, *c.App.Session(), c.App.T, "")
+	wc := c.App.NewWebConn(ws, *c.App.Session(), c.App.T, "")
+
+	if *c.App.Config().ServiceSettings.EnableReliableWebSockets {
+		connID := r.URL.Query().Get(connectionIDParam)
+		if connID == "" {
+			// If not present, we assume client is not capable yet, or it's a fresh connection.
+			// We just create a new ID.
+			connID = model.NewId()
+		} else {
+			if !model.IsValidId(connID) {
+				mlog.Error("Invalid connection ID", mlog.String("id", connID))
```

Comment:
Gentle reminder to check our logging [principles](https://developers.mattermost.com/contribute/server/style-guide/#log-levels) before merging this change.